Source/WebCore/rendering/RenderLayerFilters.cpp

```diff
@@ -185,44 +172,34 @@ GraphicsContext* RenderLayerFilters::beginFilterEffect(RenderElement& renderer,
 
     if (!filter.hasFilterThatMovesPixels())
         m_repaintRect = dirtyRect;
+    else if (hasUpdatedBackingStore)
+        m_repaintRect = filterRegion;
     else {
-        if (hasUpdatedBackingStore)
-            m_repaintRect = filterRegion;
-        else {
-            m_repaintRect = dirtyRect;
-            m_repaintRect.unite(layerRepaintRect);
-            m_repaintRect.intersect(filterRegion);
-        }
+        m_repaintRect = dirtyRect;
+        m_repaintRect.unite(layerRepaintRect);
+        m_repaintRect.intersect(filterRegion);
     }
 
-    m_paintOffset = filterRegion.location();
     resetDirtySourceRect();
-
     filter.setFilterRegion(m_filterRegion);
-    allocateBackingStoreIfNeeded(context);
 
-    auto* sourceGraphicsContext = inputContext();
-    if (!sourceGraphicsContext)
+    if (!m_targetSwitcher || hasUpdatedBackingStore)
+        m_targetSwitcher = FilterTargetSwitcher::create(context, filter, filterRegion, DestinationColorSpace::SRGB());
```

Comment:
Do we need to heap allocate this, or can it live on the stack in the caller, and get passed in? That way the lifetime is much clearer.

Reply:
The plan is to add a superclass of FilterTargetSwitcher named FilterStyleTargetSwitcher for applying the filters using GraphicsContext filters. So yes we need to keep this heap allocated.